src/coreclr/jit/llvm.cpp

```diff
@@ -2392,6 +2451,64 @@ void Llvm::PlaceAndConvertShadowStackLocals()
     lowerToShadowStack();
 }
 
+static bool placeLocalInAlloca(LclVarDsc* varDsc)
+{
+    return !varDsc->lvTracked || varDsc->lvHasLocalAddr;
+}
+
+void Llvm::createAllocasForLocalsWithAddrOp()
+{
+    m_allocas = new std::vector<Value*>(_compiler->lvaCount, nullptr);
+
+    for (unsigned lclNum = 0; lclNum < _compiler->lvaCount; lclNum++)
+    {
+        LclVarDsc* varDsc = _compiler->lvaGetDesc(lclNum);
+        if (varDsc->lvIsParam && placeLocalInAlloca(varDsc))
+        {
+            // TODO-LLVM : copy param to alloca storage
+            // see S_P_CoreLib_System_Diagnostics_Tracing_EventSource__InitializeIsSupported
+            failFunctionCompilation();
+        }
+
+        // TODO-LLVM LCLBLK - outgoing arg space
+        if (varDsc->lvType != TYP_LCLBLK && canStoreLocalOnLlvmStack(varDsc) && placeLocalInAlloca(varDsc))
+        {
+            CORINFO_CLASS_HANDLE classHandle = NO_CLASS_HANDLE;
+            if (varDsc->lvType == TYP_STRUCT)
+            {
+                ClassLayout* layout = varDsc->GetLayout();
+                classHandle         = layout->GetClassHandle();
+            }
+            Type* llvmType = getLlvmTypeForCorInfoType(toCorInfoType(varDsc->lvType), classHandle);
+            if (llvmType->isIntegerTy() && llvmType->getIntegerBitWidth() < 32)
+            {
+                llvmType = Type::getInt32Ty(_llvmContext);
+            }
+            (*m_allocas)[lclNum] = _prologBuilder.CreateAlloca(llvmType);
+        }
+    }
+}
+
+void Llvm::DontTrackLocalsWithAddress()
+{
+    for (BasicBlock* _currentBlock : _compiler->Blocks())
+    {
+        _currentRange = &LIR::AsRange(_currentBlock);
+        for (GenTree* node : CurrentRange())
+        {
+            if (node->OperIsLocalAddr())
+            {
+                LclVarDsc* localVarDsc = _compiler->lvaGetDesc(node->AsLclVarCommon()->GetLclNum());
+                if (localVarDsc->lvTracked)
+                {
+                    assert(false);// TODO-LLVM this is never hit?
+                    // localVarDsc->lvTracked = 0;
+                }
```

Comment:
I would expect you can hit this with a struct argument:
```cs
void Caller(Struct s)
{
    Callee(s);
}
```

Reply:
Thanks, I'll do the `lvIsParam` work and then that should get it.

Reply:
I added this test:

```cs
    struct LlvmStruct
    {
        internal int i1, i2;
    }

    static int FuncWithStructArg(LlvmStruct s)
    {
        return s.i2;
    }

    private static void TestStructTracking(LlvmStruct s)
    {
        StartTest("Test passing struct arg");
        
        EndTest(FuncWithStructArg(s) == 2);
    }
```
The LIR ends up (up to the CALL):
```
------------ BB01 [???..???), preds={} succs={BB02}

------------ BB02 [000..01D) (return), preds={BB01} succs={}
               [000024] ------------                 IL_OFFSET void   IL offset: 0x0
               [000025] ------------                 IL_OFFSET void   IL offset: 0x1
N002 (  1,  4) [000019] H-----------        t19 =    CNS_INT(h) ref    0x420098 [ICON_STR_HDL]
               [000031] ------------        t31 =    LCL_VAR   int    V03 tmp2         u:1
               [000032] ------------        t32 =    CNS_INT   int    0
                                                  /--*  t31    int
                                                  +--*  t32    int
               [000033] ------------        t33 = *  ADD       int
                                                  /--*  t33    int
               [000034] ------------        t34 = *  PUTARG_TYPE int
               [000035] ------------        t35 =    LCL_VAR   int    V03 tmp2         u:1
               [000036] ------------        t36 =    CNS_INT   int    0
                                                  /--*  t35    int
                                                  +--*  t36    int
               [000037] ------------        t37 = *  ADD       int
                                                  /--*  t37    int
                                                  +--*  t19    ref
               [000038] ------------              *  STOREIND  ref
                                                  /--*  t34    int    arg-1
N003 ( 15,  7) [000003] --CXG-------              *  CALL      void   Program.StartTest
               [000026] ------------                 IL_OFFSET void   IL offset: 0xb
               [000027] ------------                 IL_OFFSET void   IL offset: 0xc
N002 (  3,  2) [000005] -------N----         t5 =    LCL_VAR_ADDR byref  V00 arg0          (last use)
                                                  /--*  t5     byref
N004 (  9,  7) [000008] n-----------         t8 = *  OBJ       struct<Program+LlvmStruct, 8>
               [000039] ------------        t39 =    LCL_VAR   int    V03 tmp2         u:1
               [000040] ------------        t40 =    CNS_INT   int    0
                                                  /--*  t39    int
                                                  +--*  t40    int
               [000041] ------------        t41 = *  ADD       int
                                                  /--*  t41    int
               [000042] ------------        t42 = *  PUTARG_TYPE int
                                                  /--*  t8     struct
               [000043] ------------        t43 = *  PUTARG_TYPE struct
                                                  /--*  t42    int    arg-1
                                                  +--*  t43    struct arg0
N005 ( 23, 10) [000006] --CXG-------         t6 = *  CALL      int    Program.FuncWithStructArg
```

Resulting in LLVM:
```
define void @HelloWasm_Program__TestStructTracking(i8* %0, %"[HelloWasm]Program+LlvmStruct" %1) !dbg !5 {
Prolog:
  %2 = alloca %"[HelloWasm]Program+LlvmStruct", align 8
  store %"[HelloWasm]Program+LlvmStruct" %1, %"[HelloWasm]Program+LlvmStruct"* %2, align 1
  br label %3

3:                                                ; preds = %Prolog
  br label %4

4:                                                ; preds = %3
  %5 = load i32*, i32** @__Str_Test_passing_struct_arg_402F9E1D2A2D7F7DD38FC1B0EB39F0651A66125099651830DE99655CE79A453E___SYMBOL, align 4, !dbg !7
  %6 = getelementptr i8, i8* %0, i32 0, !dbg !7
  %7 = getelementptr i8, i8* %0, i32 0, !dbg !7
  %8 = bitcast i8* %7 to i8**, !dbg !7
  %9 = bitcast i32* %5 to i8*, !dbg !7
  store i8* %9, i8** %8, align 4, !dbg !7
  call void @HelloWasm_Program__StartTest(i8* %6), !dbg !7
  %10 = load %"[HelloWasm]Program+LlvmStruct", %"[HelloWasm]Program+LlvmStruct"* %2, align 1, !dbg !8
  %11 = getelementptr i8, i8* %0, i32 0, !dbg !8
  %12 = call i32 @HelloWasm_Program__FuncWithStructArg(i8* %11, %"[HelloWasm]Program+LlvmStruct" %10), !dbg !8
  %13 = icmp eq i32 %12, 2, !dbg !9
  %14 = getelementptr i8, i8* %0, i32 0, !dbg !9
  %15 = getelementptr i8, i8* %0, i32 0, !dbg !9
  %16 = bitcast i8* %15 to i8**, !dbg !9
  store i8* null, i8** %16, align 4, !dbg !9
  %17 = zext i1 %13 to i8, !dbg !9
  call void @HelloWasm_Program__EndTest(i8* %14, i8 %17), !dbg !9
  ret void, !dbg !9
}
```
but that assert is not hit.  We do enter `DontTrackLocalsWithAddress` for `V00` but `lvTracked` is `0`

```
<!DOCTYPE html>
<HTML>
<head>
<title>Document</title></head>
<body>
<!--StartFragment-->


  | Name | Value | Type
-- | -- | -- | --
▶ | this | 0x000002164fe40420 {_compiler=0x000002164d44e058 [Program:TestStructTracking(Program+LlvmStruct) (0x2993ab77)] ...} | Llvm *
▶ | _currentBlock | 0x000002164d44f670 BB2; BBJ_RETURN | BasicBlock *
▶ | <begin>$L0 | {m_block=0x000002164d44f670 BB2; BBJ_RETURN } | BasicBlockIterator
▶ | <begin>$L1 | {m_node=0x000002164d44fe58 [GT_LCL_VAR_ADDR, TYP_BYREF V0] } | LIR::ReadOnlyRange::Iterator
▶ | <end>$L0 | {m_block=0x0000000000000000 <NULL> } | BasicBlockIterator
▶ | <end>$L1 | {m_node=0x0000000000000000 <NULL> } | LIR::ReadOnlyRange::Iterator
▶ | <range>$L0 | {m_begin=0x000002164d44f3d0 BB1; BBJ_NONE } | BasicBlockSimpleList &&
▶ | <range>$L1 | {...} | LIR::Range &
◢ | localVarDsc | 0x000002164d44ead0 [V0: TYP_STRUCT] | LclVarDsc *
  | lvType | TYP_STRUCT (15 '\xf') | var_types
  | lvIsParam | 1 '\x1' | unsigned char
  | lvIsRegArg | 1 '\x1' | unsigned char
  | lvFramePointerBased | 0 '\0' | unsigned char
  | lvOnFrame | 1 '\x1' | unsigned char
  | lvRegister | 0 '\0' | unsigned char
  | lvTracked | 0 '\0' | unsigned char
  | lvPinned | 0 '\0' | unsigned char
  | lvMustInit | 0 '\0' | unsigned char
  | lvAddrExposed | 0 '\0' | unsigned char
  | lvDoNotEnregister | 1 '\x1' | unsigned char
  | lvFieldAccessed | 0 '\0' | unsigned char
  | lvLiveInOutOfHndlr | 0 '\0' | unsigned char
  | lvInSsa | 0 '\0' | unsigned char
  | lvVMNeedsStackAddr | 0 '\0' | unsigned char
  | lvLclFieldExpr | 0 '\0' | unsigned char
  | lvLclBlockOpAddr | 0 '\0' | unsigned char
  | lvLiveAcrossUCall | 0 '\0' | unsigned char
  | lvIsCSE | 0 '\0' | unsigned char
  | lvHasLdAddrOp | 0 '\0' | unsigned char
  | lvStackByref | 0 '\0' | unsigned char
  | lvHasILStoreOp | 0 '\0' | unsigned char
  | lvHasMultipleILStoreOp | 0 '\0' | unsigned char
  | lvIsTemp | 0 '\0' | unsigned char
  | lvIsBoolean | 0 '\0' | unsigned char
  | lvSingleDef | 1 '\x1' | unsigned char
  | lvSingleDefRegCandidate | 1 '\x1' | unsigned char
  | lvDisqualifySingleDefRegCandidate | 0 '\0' | unsigned char
  | lvSpillAtSingleDef | 0 '\0' | unsigned char
  | lvDisqualify | 0 '\0' | unsigned char
  | lvVolatileHint | 0 '\0' | unsigned char
  | lvStructDoubleAlign | 0 '\0' | unsigned char
  | lvKeepType | 0 '\0' | unsigned char
  | lvNoLclFldStress | 0 '\0' | unsigned char
  | lvIsPtr | 0 '\0' | unsigned char
  | lvIsUnsafeBuffer | 0 '\0' | unsigned char
  | lvPromoted | 0 '\0' | unsigned char
  | lvIsStructField | 0 '\0' | unsigned char
  | lvOverlappingFields | 0 '\0' | unsigned char
  | lvContainsHoles | 0 '\0' | unsigned char
  | lvCustomLayout | 0 '\0' | unsigned char
  | lvIsMultiRegArg | 0 '\0' | unsigned char
  | lvIsMultiRegRet | 0 '\0' | unsigned char
  | lvUnusedStruct | 0 '\0' | unsigned char
  | lvUndoneStructPromotion | 0 '\0' | unsigned char
  | lvLRACandidate | 0 '\0' | unsigned char
  | lvRegStruct | 0 '\0' | unsigned char
  | lvClassIsExact | 0 '\0' | unsigned char
  | lvClassInfoUpdated | 0 '\0' | unsigned char
  | lvImplicitlyReferenced | 1 '\x1' | unsigned char
  | lvSuppressedZeroInit | 0 '\0' | unsigned char
  | lvHasExplicitInit | 0 '\0' | unsigned char
  | lvFieldLclStart | 0 | unsigned int
  | lvParentLcl | 0 | unsigned int
  | lvFieldCnt | 0 '\0' | unsigned char
  | lvFldOffset | 0 '\0' | unsigned char
  | lvFldOrdinal | 0 '\0' | unsigned char
  | lvLlvmArgNum | 1 | unsigned int
  | lvCorInfoType | CORINFO_TYPE_VALUECLASS (19) | CorInfoType
  | lvHasLocalAddr | 1 '\x1' | unsigned char
  | lvSingleDefDisqualifyReason | 72 'H' | unsigned char
  | _lvRegNum | 0 '\0' | unsigned char
  | _lvOtherReg | 0 '\0' | unsigned char
  | _lvArgReg | 1 '\x1' | unsigned char
  | _lvArgInitReg | 0 '\0' | unsigned char
  | lvVarIndex | 0 | unsigned short
  | m_lvRefCnt | 4 | unsigned short
  | m_lvRefCntWtd | 400.000000 | float
  | lvStkOffs | -1163005939 | int
  | lvExactSize | 8 | unsigned int
  | lvSlotNum | 0 | unsigned int
  | ▶ lvVerTypeInfo | {m_bits={type=TI_STRUCT (2) uninitobj=0 byref=0 ...} m_flags=2 m_cls=0x4000000000420028 {...} ...} | typeInfo
  | lvClassHnd | 0x4000000000420028 {...} | CORINFO_CLASS_STRUCT_ *
  | lvFieldHnd | 0x0000000000000000 <NULL> | CORINFO_FIELD_STRUCT_ *
  | ▶ m_layout | 0x000002164d44f370 {m_classHandle=0x4000000000420028 {...} m_size=8 m_isValueClass=1 ...} | ClassLayout *
  | ▶ lvRefBlks | 0x0000000000000000 {???} | unsigned __int64 *
  | ▶ lvDefStmt | 0x0000000000000000 <NULL> | Statement *
  | ▶ lvPerSsaData | {m_array=0x0000000000000000 <NULL> m_arraySize=0 m_count=0 } | SsaDefArray<LclSsaVarDsc>
  | ▶ lvReason | 0x0000000000000000 <NULL> | const char *
▶ | node | 0x000002164d44fe58 [GT_LCL_VAR_ADDR, TYP_BYREF V0] | GenTree * {GenTreeLclVar}


<!--EndFragment-->
</body>
</HTML>
```

Reply:
Yes...

I realized it might not be that straightforward to clear `lvTracked` before SSA. The ordering here is:
```c
fgSsaBuild:
  lvaSortByRefCount();  // Sets lvTracked
  fgLocalVarLiveness(); // We'd like it to use the ordinary definition for "lvTracked"
  <SSA construction>    // We'd like it to not include address-taken variables here
```
What we can do (and I think is the right thing to do) is modify `IncludeInSsa` s.t. that ignore address-taken variables.

Then the test for whether a local should be in an `alloca` is `!lvaInSsa(lclNum)`.